**This tutorial is from Prof. Xin Tong's DSO 530 class at USC.**

## 1. Supplementary Part of Multiple Linear Regression

In [2]:
import pandas as pd
housing = pd.read_csv("Housing.csv")
display(housing.head())

crim    zn  river     rm  ptratio  medv
0  0.00632  18.0      0  6.575     15.3  24.0
1  0.02731   0.0      0  6.421     17.8  21.6
2  0.02729   0.0      0  7.185     17.8  34.7
3  0.03237   0.0      0  6.998     18.7  33.4
4  0.06905   0.0      0  7.147     18.7  36.2

### 1.1 Simple Linear Regression

Use $smf.ols()$ function to fit a simple linear regression model with $medv$ as the response and $crim$ as the predictor. 

The basic syntax is: **smf.ols('y ~ x', data)**, where $y$ is the response, $x$ is the predictor, and $data$ is the data set in which these 2 variables are kept.

**$smf.ols()$ function only takes in data as pandas dataframes, not numpy array**

In [3]:
import statsmodels.formula.api as smf

result1 = smf.ols('medv~crim', data=housing).fit()

In [4]:
# output detailed information about the model
result1.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                   medv   R-squared:                       0.151
Model:                            OLS   Adj. R-squared:                  0.149
Method:                 Least Squares   F-statistic:                     89.49
Date:                Mon, 01 Mar 2021   Prob (F-statistic):           1.17e-19
Time:                        00:25:42   Log-Likelihood:                -1798.9
No. Observations:                 506   AIC:                             3602.
Df Residuals:                     504   BIC:                             3610.
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept     24.0331      0.409     58.740      0.000      23.229      24.837
crim          -0.4152      0.044     -9.460      0.000      -0.501      -0.329
==============================================================================
Omnibus:                      139.832   Durbin-Watson:                   0.713
Prob(Omnibus):                  0.000   Jarque-Bera (JB):              295.404
Skew:                           1.490   Prob(JB):                     7.14e-65
Kurtosis:                       5.264   Cond. No.                         10.1
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

### 1.2 Multiple Regression

To fit a multiple linear regression model using least squares, use $smf.ol()$ function. 

The syntax: **smf.ols('y ~ x1+x2+x3', data)** is used to fit a model with 3 predictors x1,x2,x3. 

In [5]:
result2 = smf.ols('medv ~crim+rm', data=housing).fit()
result2.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                   medv   R-squared:                       0.542
Model:                            OLS   Adj. R-squared:                  0.540
Method:                 Least Squares   F-statistic:                     297.6
Date:                Mon, 01 Mar 2021   Prob (F-statistic):           5.22e-86
Time:                        00:25:43   Log-Likelihood:                -1642.7
No. Observations:                 506   AIC:                             3291.
Df Residuals:                     503   BIC:                             3304.
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept    -29.2447      2.588    -11.300      0.000     -34.330     -24.160
crim          -0.2649      0.033     -8.011      0.000      -0.330      -0.200
rm             8.3911      0.405     20.726      0.000       7.596       9.186
==============================================================================
Omnibus:                      172.412   Durbin-Watson:                   0.807
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             1047.536
Skew:                           1.349   Prob(JB):                    3.39e-228
Kurtosis:                       9.512   Cond. No.                         92.3
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

Since the housing data set has 5 covariates, it would be cumbersome to type all of them. Instead we can use a short-hand:

In [6]:
# get all the variable with correct format, which would be used
# in smf.ols except for the target medv
string_cols = ' + '.join(housing.columns[:-1])

# 'string'.format() is a string formatting method in Python
# allows multiple substitutions and value formatting
# lets us concatenate elements within a string through positional formatting
result3 = smf.ols('medv ~ {}'.format(string_cols), data=housing).fit()
result3.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                   medv   R-squared:                       0.605
Model:                            OLS   Adj. R-squared:                  0.601
Method:                 Least Squares   F-statistic:                     153.3
Date:                Mon, 01 Mar 2021   Prob (F-statistic):           1.76e-98
Time:                        00:25:44   Log-Likelihood:                -1605.1
No. Observations:                 506   AIC:                             3222.
Df Residuals:                     500   BIC:                             3248.
Df Model:                           5                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept     -4.8294      4.014     -1.203      0.230     -12.716       3.057
crim          -0.1981      0.032     -6.237      0.000      -0.260      -0.136
zn             0.0277      0.012      2.230      0.026       0.003       0.052
river          3.3018      1.033      3.196      0.001       1.272       5.332
rm             7.1443      0.405     17.627      0.000       6.348       7.941
ptratio       -0.9409      0.138     -6.800      0.000      -1.213      -0.669
==============================================================================
Omnibus:                      220.975   Durbin-Watson:                   0.883
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             1796.516
Skew:                           1.703   Prob(JB):                         0.00
Kurtosis:                      11.580   Cond. No.                         436.
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

In [7]:
print(string_cols)

crim + zn + river + rm + ptratio


But what if we want to perform a regression using all the variables except one? For instance, in the above output, $zn$ has the highest p-value (but it's small enough). 

The following syntax will perform a regression using all the predictors except $zn$:

In [8]:
# Use $difference$ to exclude $zn$. The function $difference()$ returns a set 
# with the difference between 2 sets 
# For example, if A = {100,60} and B = {60,20}.
# Then A.difference(B)={100} and B.difference(A)={20}.
# A.difference(B): from A, take out anything that B has
# B.difference(A): from B, take out anything that A has
# *Returned set contains items that only exist in the first set (exclude both)*
# Think of a Venn Diagram 

string_cols = ' + '.join(housing.columns[:-1].difference(['zn']))

result4 = smf.ols('medv ~ {}'.format(string_cols),data=housing).fit()
result4.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                   medv   R-squared:                       0.601
Model:                            OLS   Adj. R-squared:                  0.598
Method:                 Least Squares   F-statistic:                     188.9
Date:                Mon, 01 Mar 2021   Prob (F-statistic):           1.42e-98
Time:                        00:25:45   Log-Likelihood:                -1607.6
No. Observations:                 506   AIC:                             3225.
Df Residuals:                     501   BIC:                             3246.
Df Model:                           4                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept     -3.8668      4.007     -0.965      0.335     -11.739       4.005
crim          -0.2035      0.032     -6.402      0.000      -0.266      -0.141
ptratio       -1.0345      0.132     -7.816      0.000      -1.295      -0.774
river          3.0411      1.031      2.951      0.003       1.016       5.066
rm             7.3223      0.399     18.354      0.000       6.538       8.106
==============================================================================
Omnibus:                      215.939   Durbin-Watson:                   0.887
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             1751.602
Skew:                           1.655   Prob(JB):                         0.00
Kurtosis:                      11.493   Cond. No.                         311.
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

In [9]:
housing.columns

Index(['crim', 'zn', 'river', 'rm', 'ptratio', 'medv'], dtype='object')

In [10]:
# exclude medv column
# .difference['zn']: only include the items in the first set (not including those in both sets)
# both set has 'zn'. Thus, 'zn' is not included in returned set.
string_cols = ' + '.join(housing.columns[:-1].difference(['zn']))
print(string_cols)

crim + ptratio + river + rm


### 1.3 Interaction Terms

$x1:x2$ tells Python to include an interaction term between $x1$ and $x2$. 

The syntax $river:rm$ includes $river, rm$, and the interaction term $riverxrm$ as predictors. Aka, $river+rm+river:rm$

In [11]:
result4 = smf.ols('medv ~ river*rm', data=housing).fit()
# this is the same as:
# result4 = smf.ols('medv ~ river+rm+rvier:rm', data=housing).fit()

result4.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                   medv   R-squared:                       0.496
Model:                            OLS   Adj. R-squared:                  0.493
Method:                 Least Squares   F-statistic:                     164.9
Date:                Mon, 01 Mar 2021   Prob (F-statistic):           2.24e-74
Time:                        00:25:46   Log-Likelihood:                -1666.7
No. Observations:                 506   AIC:                             3341.
Df Residuals:                     502   BIC:                             3358.
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept    -34.4616      2.776    -12.415      0.000     -39.915     -29.008
river          7.5633      8.871      0.853      0.394      -9.865      24.992
rm             9.0241      0.440     20.496      0.000       8.159       9.889
river:rm      -0.5361      1.355     -0.396      0.692      -3.198       2.125
==============================================================================
Omnibus:                       93.496   Durbin-Watson:                   0.748
Prob(Omnibus):                  0.000   Jarque-Bera (JB):              564.362
Skew:                           0.638   Prob(JB):                    2.82e-123
Kurtosis:                       8.014   Cond. No.                         199.
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

### 1.4 Non-linear Transformations of the Predictors

Given a predictor $X$, we can create a predictor $X^2$ using $np.power(X,2)$.

Now we can perform a regression of $medv$ onto $rm$ and $rm^2$:

In [12]:
import numpy as np
result5 = smf.ols('medv ~ rm+np.power(rm,2)', data=housing).fit()
result5.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                   medv   R-squared:                       0.548
Model:                            OLS   Adj. R-squared:                  0.547
Method:                 Least Squares   F-statistic:                     305.4
Date:                Mon, 01 Mar 2021   Prob (F-statistic):           1.46e-87
Time:                        00:25:47   Log-Likelihood:                -1639.1
No. Observations:                 506   AIC:                             3284.
Df Residuals:                     503   BIC:                             3297.
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
===================================================================================
                      coef    std err          t      P>|t|      [0.025      0.975]
-----------------------------------------------------------------------------------
Intercept          66.0588     12.104      5.458      0.000      42.278      89.839
rm                -22.6433      3.754     -6.031      0.000     -30.019     -15.267
np.power(rm, 2)     2.4701      0.291      8.502      0.000       1.899       3.041
==============================================================================
Omnibus:                       82.173   Durbin-Watson:                   0.689
Prob(Omnibus):                  0.000   Jarque-Bera (JB):              934.337
Skew:                           0.224   Prob(JB):                    1.29e-203
Kurtosis:                       9.642   Cond. No.                     1.91e+03
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 1.91e+03. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

### 1.5 Confidence Interval and Prediction Interval

- **Confidence Interval**: average response
- **Prediction Interval**: particular response

In [13]:
result1 = smf.ols('medv ~ rm', data=housing).fit()
result1.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                   medv   R-squared:                       0.484
Model:                            OLS   Adj. R-squared:                  0.483
Method:                 Least Squares   F-statistic:                     471.8
Date:                Mon, 01 Mar 2021   Prob (F-statistic):           2.49e-74
Time:                        00:25:48   Log-Likelihood:                -1673.1
No. Observations:                 506   AIC:                             3350.
Df Residuals:                     504   BIC:                             3359.
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept    -34.6706      2.650    -13.084      0.000     -39.877     -29.465
rm             9.1021      0.419     21.722      0.000       8.279       9.925
==============================================================================
Omnibus:                      102.585   Durbin-Watson:                   0.684
Prob(Omnibus):                  0.000   Jarque-Bera (JB):              612.449
Skew:                           0.726   Prob(JB):                    1.02e-133
Kurtosis:                       8.190   Cond. No.                         58.4
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

Next, we can produce confidence intervals and prediction intervals for the prediction of $medv$ (y) for a given value of $rm$ (x).

Assume that we want to predict for the following given value of $rm$.

In [18]:
# set 4 predictions "x" to df
test_data = {'rm': [5,10,15,20]}
test_data_df = pd.DataFrame(test_data)
test_data_df

rm
0   5
1  10
2  15
3  20

We can use get_prediction() function to produce confidence intervals and prediction intervals for the prediction.

In [17]:
prediction1 = result1.get_prediction(test_data_df)
prediction1.summary_frame(alpha=0.05)

mean   mean_se  mean_ci_lower  mean_ci_upper  obs_ci_lower  \
0   10.839924  0.613410       9.634769      12.045079     -2.214474   
1   56.350469  1.584377      53.237672      59.463266     42.984303   
2  101.861014  3.663795      94.662822     109.059205     87.002385   
3  147.371559  5.754624     136.065553     158.677565    130.143945   

   obs_ci_upper  
0     23.894322  
1     69.716635  
2    116.719643  
3    164.599173

The 95% CI associated with a $rm$ value of 10 is (53.24, 59.46).

The 95% PI associated with a $rm$ value of 10 is (42.98, 69.72).

Both intervals are centered around the same point (predicted value is the mean 56.35), but PI is wider.

In this example, $CI = B_0 + 10B_1$